### PR TITLE
8269476: Skip nmethod entry barrier if there is no oops in the jit code

### PR DESCRIPTION
**Pull request by casparcwang**
Lots of c1 and c2 jit methods do not contain any oop, so the nmethod entry barrier can be skipped.

1, c1 jit code will patch oops or Klass into the nmethod, so the entry barrier cannot directly be eliminated, current implementation uses a jump instruction to replace the jcc instruction. If the jit code is patched to contain oops, the entry barrier is patched back to the jcc instruction.

2, only the jit code of core library methods do not contain any oops.

3, currently only support zgc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8269476](https://bugs.openjdk.java.net/browse/JDK-8269476): Skip nmethod entry barrier if there is no oops in the jit code

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4610/head:pull/4610` \
`$ git checkout pull/4610`

Update a local copy of the PR: \
`$ git checkout pull/4610` \
`$ git pull https://git.openjdk.java.net/jdk pull/4610/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4610`

View PR using the GUI difftool: \
`$ git pr show -t 4610`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4610.diff">https://git.openjdk.java.net/jdk/pull/4610.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp [hunk 1]
```diff
@@ -161,3 +161,7 @@ bool BarrierSetNMethod::is_armed(nmethod* nm) {
   NativeNMethodBarrier* barrier = native_nmethod_barrier(nm);
   return barrier->get_value() != disarmed_value();
 }
+
+void BarrierSetNMethod::fix_entry_barrier(nmethod*, bool) {
+  // not implement yet
```
Comment (DamonFool) [inline src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp:166]: Shall we use `Unimplemented();` here?

Comment (casparcwang) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: refactor to x86 only

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: I think there is a simpler way to accomplish this by playing with disarmed_value, e.g. setting nmethod's disarmed_value to certain pattern to indicate it is always disarmed.

Comment (casparcwang) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: > I think there is a simpler way to accomplish this by playing with disarmed_value, e.g. setting nmethod's disarmed_value to certain pattern to indicate it is always disarmed.

`__ cmpl(disarmed_addr, 0);` will only set ZF=1 when `dest = source`. By playing with disarmed_value, do you mean also change the cmpl instruction to other instruction?

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: You may need an additional instruction, e.g. uses MSB of disarmed_value to indicate always disarmed and test the bit before `cmpl`. 

Also, you need a new method to set the bit during nmethod registration.

nmethod entry barrier tests is_armed() early, so may benefit more.

Comment (casparcwang) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: > You may need an additional instruction, e.g. uses MSB of disarmed_value to indicate always disarmed and test the bit before `cmpl`.
> 
> Also, you need a new method to set the bit during nmethod registration.
> 
> nmethod entry barrier tests is_armed() early, so may benefit more.

The barrier can be bypassed only when the nmethod does not any contain oop relocation, if it's bypassed, it should always be bypassed, there is no need to test the bit of disarmed_value.

I think several `nop` instruction can be added before `cmpl`, and patched to direct jump if the barrier can be bypassed. And when the nmethod is changed to contain oops again, the direct jump should be patched to `nop` again. 

`x86_instruction_opcode  disarmed_value_addr, some_magic_number_immediate; jcc label;` : If there exists some instructions like this, only need to change `some_magic_number_immediate` to make the following `jcc` instruction alway jump, this is the dream version of code sequences, but I do not figure out which instructions can be used to implement this functionality.

Comment (casparcwang) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: maybe `subl ; jcc` can be used to implement a more compact version, but the `nop`<->`jump` can benefit more

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: > > You may need an additional instruction, e.g. uses MSB of disarmed_value to indicate always disarmed and test the bit before `cmpl`.
> > Also, you need a new method to set the bit during nmethod registration.
> > nmethod entry barrier tests is_armed() early, so may benefit more.
> 
> The barrier can be bypassed only when the nmethod does not any contain oop relocation, if it's bypassed, it should always be bypassed, there is no need to test the bit of disarmed_value.

On AArch64 we have
```
  __ ldrw(rscratch1, guard);

  // Subsequent loads of oops must occur after load of guard value.                                                                                                                                                                                                                    
  // BarrierSetNMethod::disarm sets guard with release semantics.                                                                                                                                                                                                                      
  __ membar(__ LoadLoad);
  __ ldrw(rscratch2, thread_disarmed_addr);
  __ cmpw(rscratch1, rscratch2);
  __ br(Assembler::EQ, skip);
```
This `membar` is evil, and it would be very nice to be rid of it. If we have a `nop` before the guard load, we can patch it to jump around the whole sequence, so that would be my preferred thing to do.

On AArch64 we don't care about C1 patching at all, we just don't do any. It would be better IMO if everyone gave up C1 patching, but that's a discussion for another day.

**Comment by zhengyu123** [comment]
nmethod entry barrier is inserted in SharedRuntime::generate_native_wrapper(). 

At that point, the method is already compiled and its metadata information should be available. It should be possible to determine if there are any embedded oops there and elide barrier completely if there is none.

**Comment by zhengyu123** [comment]
Following code should help you to find out if the method contains any relocatable oops, so you can generate nmethod entry barrier accordingly ...

`diff --git a/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp b/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp`
`index fb4e3f54400..4c2686d27db 100644`
`--- a/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp`
`+++ b/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp`
`@@ -1803,6 +1803,9 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,`
`   // -2 because return address is already present and so is saved rbp
`   __ subptr(rsp, stack_size - 2*wordSize);`
 
`+  bool can_elide_nmethod_barrier = !nmethod::has_relocatable_oops(masm->code());`
`+  // This nmethod has no relocatable oop, can elide the barrier.`
`+  // However, we still need to generate something to not crash nmethod arm/disarm calls.`
`   BarrierSetAssembler* bs = BarrierSet::barrier_set()->barrier_set_assembler();`
`   bs->nmethod_entry_barrier(masm);`
 
`diff --git a/src/hotspot/share/code/nmethod.cpp b/src/hotspot/share/code/nmethod.cpp`
`index e2b27e5f4f0..99c7276e952 100644`
`--- a/src/hotspot/share/code/nmethod.cpp`
`+++ b/src/hotspot/share/code/nmethod.cpp`
`@@ -879,6 +879,19 @@ void nmethod::log_identity(xmlStream* log) const {`
` #endif`
` }`
 
`+bool nmethod::has_relocatable_oops(const CodeBuffer* cb) {`
`+  for (int n = (int) CodeBuffer::SECT_FIRST; n < (int)CodeBuffer::SECT_LIMIT; n++) {`
`+    const CodeSection* cs = cb->code_section(n);`
`+    RelocIterator iter(const_cast<CodeSection*>(cs));`
`+    while (iter.next()) {`
`+      if (iter.type() == relocInfo::oop_type) {`
`+        // Found relocatable oop`
`+        return true;`
`+      }`
`+    }`
`+  }`
`+  return false;`
`+}`
 
 `#define LOG_OFFSET(log, name)                    \`
`   if (p2i(name##_end()) - p2i(name##_begin())) \`
`diff --git a/src/hotspot/share/code/nmethod.hpp b/src/hotspot/share/code/nmethod.hpp`
`index 893f28863a6..1504ae9bc51 100644`
`--- a/src/hotspot/share/code/nmethod.hpp`
`+++ b/src/hotspot/share/code/nmethod.hpp`
`@@ -555,6 +555,8 @@ public:`
`   // Verify calls to dead methods have been cleaned.`
`   void verify_clean_inline_caches();`
 
`+  static bool has_relocatable_oops(const CodeBuffer* cb);`
`+`
 `  // unlink and deallocate this nmethod`
 `  // Only NMethodSweeper class is expected to use this. NMethodSweeper is not`
 `  // expected to use any other private methods/data in this class.`

File: src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp (hunk 1 above)
Comment (casparcwang) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp]: > This `membar` is evil, and it would be very nice to be rid of it. If we have a `nop` before the guard load, we can patch it to jump around the whole sequence, so that would be my preferred thing to do.

thanks for your suggestion, I'll try to implement the `nop` version on aarch64 architecture.

**Comment by casparcwang** [comment]
> nmethod entry barrier is inserted in SharedRuntime::generate_native_wrapper().
> 
> At that point, the method is already compiled and its metadata information should be available. It should be possible to determine if there are any embedded oops there and elide barrier completely if there is none.

Thanks for providing the patch. But we cannot aways figure out there are any oops on the generation of entry barrier:
1, C1 patching mechanism will dynamically introduce java mirror oop to the nmethod, if the patching code is executed. If the oops is added again to the method, the entry barrier should be activated again.
2, In C2 compilation, `MachPrologNode::emit` will call `MacroAssembler::verified_entry` to generate the nmethod entry barrier, that's the very first node to emit machine instructions, so at this time how many oops embedded is unknown.
3, In `nmethod::new_nmethod`, it will call `CodeBuffer::finalize_oop_references` to add class loader of embedded Klass meta to the oop recorder.  Only visit relocation will miss these classloaders.

So the entry barrier can not be easily eliminated on current code base, and if the barrier needs to be eliminated, there are lots of things to do.

**Comment by fisk** [comment]
1) When this was introduced, we did not see the overhead of nmethod entry barriers in performance profiles. Did you see any improvement with the patch?
2) Any method that can be on-stack has at least an oop for the method holder, to prevent the method from being unloaded.
3) These barriers are like a swiss army knife. They do not only protect against oops in machine code. As mentioned they also protect against the class being unloaded, and racing between unloading and loading. In project loom they are also used for the nmethod lifecycle, and in generational ZGC they are used to patch the barrier code itself.
4) Having special nmethods that don't quack the same way makes the GC code unintuitive to work with and understand.

In summary, I don't think we should be eliding any nmethod entry barriers.

**Review by fisk — CHANGES_REQUESTED** [review]
I don't think we should do this at all.

**Comment by theRealAph** [comment]
On 6/29/21 8:03 AM, Erik Österlund wrote:
> 1) When this was introduced, we did not see the overhead of nmethod entry barriers in performance profiles. Did you see any improvement with the patch?

This is a LoadLoad fence at the start of every method followed by a load
with a dependent branch. If there wasn't a significant hit in the profiles
I wouldn't believe the profiles. Even adding code that doesn't appear to slow
things down uses additional speculation resources.

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

**Comment by fisk** [comment]
> On 6/29/21 8:03 AM, Erik Österlund wrote:
> > 1) When this was introduced, we did not see the overhead of nmethod entry barriers in performance profiles. Did you see any improvement with the patch?
> 
> This is a LoadLoad fence at the start of every method followed by a load
> with a dependent branch. If there wasn't a significant hit in the profiles
> I wouldn't believe the profiles. Even adding code that doesn't appear to slow
> things down uses additional speculation resources.

My main concern is that I am not convinced that eliding nmethod entry barriers is sound. It really isn't all about protecting oops found in the machine code as I explained previously. There is a lot more to it. Therefore I would like to at least have some convincing number showing that such an optimization effort is worthwhile.

If you are looking at ways of getting rid of the loadload, I have a scheme that can elide it if you are interested. I discussed it with Stuart when he was prototyping how to build these barriers, but I think he concluded we didn't need such optimizations in the end. It came with some more complexity. That's why I am curious if you have numbers suggesting the opposite. Because if so, I have a better idea to target the loadload in particular, witout violating our invariants and still being correct.

**Comment by casparcwang** [comment]
In summary, the benefit is too small, and it adds maintenance burden to future loom and generational zgc design, so just close this pr.

**Comment by theRealAph** [comment]
On 6/29/21 2:16 PM, Erik Österlund wrote:
>> On 6/29/21 8:03 AM, Erik Österlund wrote:
>>> 1) When this was introduced, we did not see the overhead of nmethod entry barriers in performance profiles. Did you see any improvement with the patch?
>>
>> This is a LoadLoad fence at the start of every method followed by a load
>> with a dependent branch. If there wasn't a significant hit in the profiles
>> I wouldn't believe the profiles. Even adding code that doesn't appear to slow
>> things down uses additional speculation resources.
> 
> My main concern is that I am not convinced that eliding nmethod entry barriers is sound. It really isn't all about protecting oops found in the machine code as I explained previously. There is a lot more to it. Therefore I would like to at least have some convincing number showing that such an optimization effort is worthwhile.
> 
> If you are looking at ways of getting rid of the loadload, I have a scheme that can elide it if you are interested. I discussed it with Stuart when he was prototyping how to build these barriers, but I think he concluded we didn't need such optimizations in the end. It came with some more complexity. That's why I am curious if you have numbers suggesting the opposite. Because if so, I have a better idea to target the loadload in particular, witout violating our invariants and still being correct.

No, I don't have numbers, and it would be a moderately serious effort to get them.
I could do it, but it seems like a ridiculous effort to eliminate something that
obviously should be eliminated.

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671
